pkg/suggestion/v1alpha3/goptuna/service.go

```diff
@@ -0,0 +1,182 @@
+package suggestion_goptuna_v1alpha3
+
+import (
+	"context"
+	"sync"
+
+	"github.com/c-bata/goptuna"
+	"github.com/kubeflow/katib/pkg/apis/manager/v1alpha3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog"
+)
+
+const (
+	AlgorithmCMAES  = "cmaes"
+	AlgorithmTPE    = "tpe"
+	AlgorithmRandom = "random"
+
+	defaultStudyName = "Katib"
+)
+
+func NewSuggestionService() *SuggestionService {
+	return &SuggestionService{
+		searchSpace:  nil,
+		study:        nil,
+		trialMapping: make(map[string]int),
+	}
+}
+
+type SuggestionService struct {
+	mu           sync.Mutex
+	searchSpace  map[string]interface{}
+	study        *goptuna.Study
+	trialMapping map[string]int // Katib trial name -> Goptuna trial id
+}
+
+func (s *SuggestionService) GetSuggestions(
+	ctx context.Context,
+	req *api_v1_alpha3.GetSuggestionsRequest,
+) (*api_v1_alpha3.GetSuggestionsReply, error) {
+	if s.study == nil || s.searchSpace == nil {
+		return nil, status.Error(codes.Unavailable, "You need to call ValidateAlgorithmSettings before")
+	}
+
+	objectMetricName := req.GetExperiment().GetSpec().GetObjective().GetObjectiveMetricName()
+	trials, err := toGoptunaTrials(req.GetTrials(), objectMetricName, s.study, s.searchSpace)
+	if err != nil {
+		klog.Errorf("Failed to convert to Goptuna trials: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	err = s.syncTrials(trials)
+	if err != nil {
+		klog.Errorf("Failed to sync Goptuna trials: %s", err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	requestNumber := int(req.GetRequestNumber())
+	parameterAssignments := make([]*api_v1_alpha3.GetSuggestionsReply_ParameterAssignments, requestNumber)
+	for i := 0; i < requestNumber; i++ {
+		assignments, err := sampleNextParam(s.study, s.searchSpace)
+		if err != nil {
+			klog.Errorf("Failed to sample next param: %s", err)
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		parameterAssignments[i] = &api_v1_alpha3.GetSuggestionsReply_ParameterAssignments{
+			Assignments: assignments,
+		}
+	}
+
+	return &api_v1_alpha3.GetSuggestionsReply{
+		ParameterAssignments: parameterAssignments,
+	}, nil
+}
+
+// Sync Goptuna trials with Katib trials.
+func (s *SuggestionService) syncTrials(ktrials map[string]goptuna.FrozenTrial) (err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
```

Comment:
FYI, this lock is to protect `s.trialMapping` and all methods of `*goptuna.Study` are goroutine-safe.

Reply:
Will we have multiple goroutines here?

Reply:
Yes. When it is received multiple concurrent requests, two or more katib trials might be mapping to the same goptuna trial id if goptuna samples the same parameter set. (e.g. Goptuna CMA-ES might sample the same parameter set in larger budgets).

The related discussion is https://github.com/kubeflow/katib/pull/1131#discussion_r407466277.